pkg/retry/retry.go

```diff
@@ -0,0 +1,143 @@
+// Package retry provides a generic retry mechanism with exponential backoff.
+//
+// Example:
+//
+//	ctx := context.Background()
+//	client := foobar.NewClient()
+//
+//	err := retry.Do(func() error {
+//		err := client.UpdateConfig(ctx, map[string]any{"key": "value"})
+//
+//		// will be retied
+//		if errors.Is(err, foobar.ErrTxConflict) {
+//			return retry.Retryable(err)
+//		}
+//
+//		return err
+//	})
+package retry
+
+import (
+	"context"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+	"github.com/pkg/errors"
+)
+
+type Backoff = backoff.BackOff
+
+type Callback func() error
+
+type TypedCallback[T any] func() (T, error)
+
+type errRetryable struct {
+	error
+}
+
+// DefaultBackoff returns a default backoff strategy with 5 exponential retries.
+func DefaultBackoff() Backoff {
+	bo := backoff.NewExponentialBackOff(
+		backoff.WithInitialInterval(50*time.Millisecond),
+		backoff.WithMaxInterval(500*time.Millisecond),
+		backoff.WithMultiplier(1.8),
+	)
+
+	return backoff.WithMaxRetries(bo, 5)
+}
+
+// Do executes the callback function with the default backoff config.
+// It will retry a callback ONLY if error is retryable.
+func Do(cb Callback) error {
+	return DoWithBackoff(cb, DefaultBackoff())
+}
+
+// DoWithBackoff executes the callback function with provided backoff config.
+// It will retry a callback ONLY if error is retryable.
+func DoWithBackoff(cb Callback, bo Backoff) error {
+	for {
+		err := cb()
+		if err == nil {
+			return nil
+		}
+
+		var errRetry errRetryable
+		isRetryable := errors.As(err, &errRetry)
+		if !isRetryable {
+			return err
+		}
+
+		sleepFor := bo.NextBackOff()
+		if sleepFor == backoff.Stop {
+			return errors.Wrap(err, "retry limit exceeded")
+		}
+
+		time.Sleep(sleepFor)
```

Comment:
I'm saying this bit here specifically ignores context cancellations because this api is not context aware